File: GencomStay/Graph/GraphClient.swift
import Foundation

/// A thin client over the Microsoft Graph REST API for reading and triaging mail.
///
/// All requests attach a bearer token fetched from the `TokenProviding` source, so
/// token refresh stays in one place (AuthManager). Responses are decoded into the
/// app's normalized `EmailMessage` model rather than exposing Graph's wire format.
actor GraphClient {
    private let baseURL = URL(string: "https://graph.microsoft.com/v1.0")!
    private let tokenProvider: TokenProviding
    private let session: URLSession
    private let decoder: JSONDecoder

    init(tokenProvider: TokenProviding, session: URLSession = .shared) {
        self.tokenProvider = tokenProvider
        self.session = session
        let decoder = JSONDecoder()
        // Graph timestamps are ISO-8601 (e.g. "2026-06-04T18:13:00Z").
        decoder.dateDecodingStrategy = .iso8601
        self.decoder = decoder
    }

    /// Fetch the most recent messages from the inbox, newest first.
    func fetchInbox(top: Int = 50) async throws -> [EmailMessage] {
        var components = URLComponents(
            url: baseURL.appendingPathComponent("me/mailFolders/inbox/messages"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "$top", value: String(top)),
            URLQueryItem(name: "$orderby", value: "receivedDateTime desc"),
            URLQueryItem(name: "$select",
                         value: "id,conversationId,subject,from,toRecipients,receivedDateTime,bodyPreview,isRead,hasAttachments,importance,webLink")
        ]
        let request = try await authorizedRequest(url: components.url!)
        let (data, response) = try await session.data(for: request)
        try Self.validate(response, data: data)
        let envelope = try decoder.decode(GraphCollection<GraphMessage>.self, from: data)
        return envelope.value.map { $0.normalized() }
    }

    /// Mark a message as read on the server (mirrors a user opening it).
    func markRead(messageId: String, isRead: Bool = true) async throws {
        let url = baseURL.appendingPathComponent("me/messages/\(messageId)")
        var request = try await authorizedRequest(url: url)
        request.httpMethod = "PATCH"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["isRead": isRead])
        let (data, response) = try await session.data(for: request)
        try Self.validate(response, data: data)
    }

    // MARK: - Internals

    private func authorizedRequest(url: URL) async throws -> URLRequest {
        let token = try await tokenProvider.currentAccessToken()
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        return request
    }

    private static func validate(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else {
            throw GraphError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw GraphError.http(status: http.statusCode, body: body)
        }
    }
}

enum GraphError: LocalizedError {
    case invalidResponse
    case http(status: Int, body: String)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Unexpected response from Microsoft Graph."
        case .http(let status, let body):
            return "Graph request failed (\(status)): \(body)"
        }
    }
}

File: src/graph/graphClient.ts
import type { EmailMessage } from "../models/types";
import type { MailSource } from "../models/mailSource";
import {
  normalizeMessage,
  type GraphCollection,
  type GraphMessage,
} from "./graphTypes";

const GRAPH_BASE = "https://graph.microsoft.com/v1.0";

/**
 * Thin client over the Microsoft Graph REST API for reading and triaging mail.
 * Every request attaches a fresh bearer token from the supplied async getter, so
 * token refresh stays entirely in the auth layer.
 */
export class GraphClient implements MailSource {
  constructor(private readonly getAccessToken: () => Promise<string>) {}

  /** Fetch recent inbox messages, newest first. */
  async fetchInbox(top = 50): Promise<EmailMessage[]> {
    const select = [
      "id",
      "conversationId",
      "subject",
      "from",
      "toRecipients",
      "receivedDateTime",
      "bodyPreview",
      "isRead",
      "hasAttachments",
      "importance",
      "webLink",
    ].join(",");
    const url =
      `${GRAPH_BASE}/me/mailFolders/inbox/messages` +
      `?$top=${top}&$orderby=receivedDateTime desc&$select=${select}`;

    const data = await this.request<GraphCollection<GraphMessage>>(url);
    return data.value.map(normalizeMessage);
  }

  /** Mark a message read/unread on the server (mirrors the user opening it). */
  async markRead(messageId: string, isRead = true): Promise<void> {
    await this.request(`${GRAPH_BASE}/me/messages/${messageId}`, {
      method: "PATCH",
      body: JSON.stringify({ isRead }),
    });
  }

  /**
   * Create a reply draft in Outlook. Graph's createReply builds a draft in the
   * original thread (with the quoted message); `comment` is our text, placed above
   * the quote. The draft lands in the user's Outlook Drafts folder. Needs Mail.ReadWrite.
   */
  async saveReplyDraft(messageId: string, bodyText: string): Promise<void> {
    await this.request(`${GRAPH_BASE}/me/messages/${messageId}/createReply`, {
      method: "POST",
      body: JSON.stringify({ comment: GraphClient.toHtml(bodyText) }),
    });
  }

  /** Send a reply in the original thread in one call. Needs Mail.Send. */
  async sendReply(messageId: string, bodyText: string): Promise<void> {
    await this.request(`${GRAPH_BASE}/me/messages/${messageId}/reply`, {
      method: "POST",
      body: JSON.stringify({ comment: GraphClient.toHtml(bodyText) }),
    });
  }

  /** Escape text and preserve line breaks for the HTML mail body. */
  private static toHtml(text: string): string {
    const escaped = text
      .replace(/&/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;");
    return escaped.replace(/\r?\n/g, "<br>");
  }

  private async request<T>(url: string, init: RequestInit = {}): Promise<T> {
    const token = await this.getAccessToken();
    const res = await fetch(url, {
      ...init,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
        ...(init.headers ?? {}),
      },
    });
    // Read once: send/reply return 202 with no body, createReply 201, PATCH 200.
    const text = await res.text().catch(() => "");
    if (!res.ok) {
      throw new Error(`Graph request failed (${res.status}): ${text}`);
    }
    return (text ? JSON.parse(text) : undefined) as T;
  }
}
